x/ibc-hooks/contracts/crosschain-swaps/src/execute.rs

```diff
@@ -0,0 +1,271 @@
+    let parsed =
+        cw_utils::parse_execute_response_data(&b).map_err(|e| ContractError::FailedSwap {
+            msg: format!("failed to parse: {e}"),
+        })?;
+    let swap_response: SwapResponse = from_binary(&parsed.data.unwrap_or_default())?;
+
+    // Build an IBC packet to forward the swap.
+    let contract_addr = &swap_msg_state.contract_addr;
+    let ts = swap_msg_state.block_time.plus_seconds(PACKET_LIFETIME);
+    let config = CONFIG.load(deps.storage)?;
+
+    // If the memo is provided we want to include it in the IBC message
+    let memo: serde_cw_value::Value = if let Some(memo) = &swap_msg_state.forward_to.next_memo {
+        serde_json_wasm::from_str(&memo.to_string()).map_err(|_e| ContractError::InvalidMemo {
+            error: format!("this should be unreachable"),
+            memo: memo.to_string(),
+        })?
+    } else {
+        serde_json_wasm::from_str("{}").unwrap()
+    };
+
+    // If tracking callbacks, we want to include the callback key in the memo
+    // without otherwise modifying the provided one
+    let memo = match config.track_ibc_callbacks {
+        true => {
+            let serde_cw_value::Value::Map(mut m) = memo else { unreachable!() };
+            m.insert(
+                serde_cw_value::Value::String(CALLBACK_KEY.to_string()),
+                serde_cw_value::Value::String(contract_addr.to_string()),
+            );
+            serde_cw_value::Value::Map(m)
+        }
+        false => memo,
+    };
+
+    // Serialize the memo. If it is an empty json object, set it to ""
+    let mut memo_str =
+        serde_json_wasm::to_string(&memo).map_err(|_e| ContractError::InvalidMemo {
+            error: "could not serialize".to_string(),
+            memo: format!("{:?}", swap_msg_state.forward_to.next_memo),
+        })?;
+    if memo_str == "{}" {
+        memo_str = String::new();
+    }
+
+    // Cosmwasm's  IBCMsg::Transfer  does not support memo.
+    // To build and send the packet properly, we need to send it using stargate messages.
+    // See https://github.com/CosmWasm/cosmwasm/issues/1477
+    let ibc_transfer = MsgTransfer {
+        source_port: "transfer".to_string(),
+        source_channel: swap_msg_state.forward_to.channel.clone(),
+        token: Some(
+            Coin::new(
+                swap_response.amount.into(),
+                swap_response.token_out_denom.clone(),
+            )
+            .into(),
+        ),
+        sender: contract_addr.to_string(),
+        receiver: swap_msg_state.forward_to.receiver.clone().into(),
+        timeout_height: None,
+        timeout_timestamp: Some(ts.nanos()),
+        memo: memo_str,
+    };
+
+    // Base response
+    let response = Response::new()
+        .add_attribute("status", "ibc_message_created")
+        .add_attribute("ibc_message", format!("{:?}", ibc_transfer));
+
+    if !config.track_ibc_callbacks || swap_msg_state.forward_to.failed_delivery.is_none() {
```

Comment:
The respone flow / general expectations isn't super clear. Can we lay this out a bit more in the README, so callers can get a better sense of it?

Reply:
Perhaps just a response format section in readme? (or code doc somewhere, e.g. where its defined)